KAFKA-16524: Metrics for KIP-853

kevin-wu24 opened the pull request:
This PR implements the following gauges in KafkaRaftMetrics, BrokerServerMetrics, and ControllerServerMetrics:

**KafkaRaftMetrics**:
- `number-of-voters`: number of voters for a KRaft topic partition. All of the replicas will report this metric.
- `number-of-observers`: number of observers being tracked by the KRaft topic partition leader.
- `uncommitted-voter-change`: 1 if there is a voter change that has not been committed, 0 otherwise.

`number-of-observers` and `uncommitted-voter-change` will only be present on the active controller, since it does not make sense for other replicas to report these metrics. If `number-of-offline-voters` is implemented, it will also only be reported by the leader.

In order to make these two metrics thread-safe, `KafkaRaftMetrics` needs to be passed into `LeaderState`, and therefore `QuorumState`. This introduces a circularity since the `KafkaRaftMetrics` constructor takes in `QuorumState`. To break the circularity for now, the logic using `QuorumState` will be moved to a `KafkaRaftMetrics#initialize` method.

**BrokerServerMetrics**:
- `ignored-static-voters`: 1 if reading the voter set from the log, 0 if the voter set comes from static configuration.

**ControllerServerMetrics**:
- `IgnoredStaticVoters`: 1 if reading the voter set from the log, 0 if the voter set comes from static configuration. (same as the metric above but for controllers)
- ~`IsObserver`: 1 if the controller node is not a voter for the KRaft cluster metadata partition, 0 otherwise.~ This metric will not be implemented in this PR. Instead, the `current-state` metric in conjunction with the presence of `kafka.server.controller` metrics can be used to provide the same information.

To implement the `ignored-static-voters` metrics, this PR introduces the `ExternalKRaftMetrics` class, which allows for higher layer metrics objects to be accessible within the `raft` module. `number-of-voters` and `ignored-static-voters` will be updated in `KRaftControlRecordStateMachine#handleBatch` when a new VotersRecord is read from log.

**Testing**:
- Unit tests in `KRaftControlRecordStateMachineTest` and `KafkaRaftMetricsTest`.
- Added/modified tests in `KafkaRaftClientReconfigTest` to verify new metrics behavior.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

kevin-wu24 committed:
adding raft metrics for KIP-853

kevin-wu24 committed:
adding brokerServerMetrics and controllerServerMetrics

jsancio commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -3515,6 +3519,14 @@ public Optional<Node> voterNode(int id, ListenerName listenerName) {
         return partitionState.lastVoterSet().voterNode(id, listenerName);
     }
 
+    public int ignoredStaticVoters() {
```
Let's return a boolean and let the metrics exporter convert the boolean to a 1 or 0.

Once this value reports true it will continue to report true for the entire lifetime of the JVM/replica. Did you consider a different implementation that doesn't increase the number of public methods of `KafkaRaftClient`? Maybe we can store this value in the metrics objects and update it when handling `KRaftControlRecordStateMachine#handleBatch`.

kevin-wu24 replied:
I agree that once this metric reports true it will continue to report as such, so repeatedly calling this function is redundant. 

However, I'm not sure how to implement this without adding some kind of public method in `KafkaRaftClient` so that `BrokerServerMetrics` and `ControllerMetdataMetrics` can access this information. If we decide to store this in the `KafkaRaftMetrics` object and update in `KRaftControlRecordStateMachine#handleBatch`, we would still need a public method in `KafkaRaftClient` to make the `KafkaRaftMetrics` object visible to those aforementioned metrics objects. Maybe this metric should be at the `KafkaRaftMetrics` level instead.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -3515,6 +3519,14 @@ public Optional<Node> voterNode(int id, ListenerName listenerName) {
+        return partitionState.lastVoterSetOffset().isPresent() ? 1 : 0;
+    }
+
+    public int isObserver() {
+        return quorum.isObserver() ? 1 : 0;
+    }
```
Let's return a boolean and let the metrics exporter convert the boolean to a 1 or 0.

Same here. Did you consider a different implementation that doesn't increase the number of public methods of `KafkaRaftClient`? Maybe we don't need to implement this metric since KRaft's `current-state` metrics already reports half of this information. I need to see if there is a metrics that returns 1 when the process role includes controller.

kevin-wu24 replied:
The `current-state` metric does include "observer" as one of its potential values.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KafkaRaftMetrics.java:
```diff
@@ -215,6 +227,22 @@ public void maybeUpdateElectionLatency(long currentTimeMs) {
         }
     }
 
+    public <T> void addLeaderMetrics(LeaderState<T> leaderState) {
+        metrics.addMetric(numObserversMetricName, (Gauge<Integer>) (config, now) -> leaderState.numObservers());
+        metrics.addMetric(uncommittedVoterChangeMetricName, (Gauge<Integer>) (config, now) -> {
+            if (leaderState.addVoterHandlerState().isEmpty() && leaderState.removeVoterHandlerState().isEmpty()) {
+                return 0;
+            } else {
+                return 1;
+            }
+        });
+    }
```
This lambda are not thread safe. `LeaderState` assumes that the kraft driver thread is the only reader and writer.

jsancio reviewed (commented):
Thanks for the changes @kevin-wu24 . Quick review of  `src/main`.

kevin-wu24 committed:
code review: introduce ExternalKRaftMetrics and make ignoredStaticVoters and number-of-voters metrics be updated in the internal partition listener

kevin-wu24 committed:
code review: update number-of-observers and uncommitted-voter-change in a thread-safe manner

kevin-wu24 committed:
missed changes from previous commits

jsancio commented on metadata/src/main/java/org/apache/kafka/controller/metrics/ControllerMetadataMetrics.java:
```diff
@@ -17,7 +17,7 @@
 package org.apache.kafka.controller.metrics;
 
-import org.apache.kafka.raft.KafkaRaftClient;
+import org.apache.kafka.raft.internals.ExternalKRaftMetricIgnoredStaticVoters;
```
The `internals` package should not be imported outside the `raft` module.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KafkaRaftMetrics.java:
```diff
@@ -184,6 +141,59 @@ public KafkaRaftMetrics(Metrics metrics, String metricGrpPrefix, QuorumState sta
         );
     }
 
+    public void initialize(QuorumState state) {
```
Let's file a Jira to remove this refactor and assign it to me.

kevin-wu24 replied:
https://issues.apache.org/jira/browse/KAFKA-18395

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/ExternalKRaftMetrics.java:
```diff
@@ -0,0 +1,35 @@
+
+    public ExternalKRaftMetrics(ExternalKRaftMetricIgnoredStaticVoters brokerServerMetrics, ExternalKRaftMetricIgnoredStaticVoters controllerMetadataMetrics) {
+        this.brokerServerMetrics = brokerServerMetrics != null ? Optional.of(brokerServerMetrics) : Optional.empty();
+        this.controllerMetadataMetrics = controllerMetadataMetrics != null ? Optional.of(controllerMetadataMetrics) : Optional.empty();
```
You can reuse `Optional.ofNullable`. It does the same thing this expression does.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/ExternalKRaftMetricIgnoredStaticVoters.java:
```diff
@@ -0,0 +1,22 @@
+ * limitations under the License.
+ */
+
+package org.apache.kafka.raft.internals;
```
This is part of `raft` public interface. Move it to the `org.apache.kafka.raft` package.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/ExternalKRaftMetricIgnoredStaticVoters.java:
```diff
@@ -0,0 +1,22 @@
+
+package org.apache.kafka.raft.internals;
+
+public interface ExternalKRaftMetricIgnoredStaticVoters {
```
I suggest naming this `ExternalKRaftMetrics`.

kevin-wu24 replied:
My decision making behind this naming style was that the `ExternalKRaftMetrics` object is what stores all of the non-kafkaRaftMetrics objects as fields. The `ExternalKRaftMetricIgnoredStaticVoters` interface is used to avoid a circular dependency on either `kafka.server.main` or `kafka.metadata.main` and defines the contract for how these metrics objects are used in the raft layer (i.e. only `IgnoredStaticVoters` should be visible from the raft layer).

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/ExternalKRaftMetrics.java:
```diff
@@ -0,0 +1,35 @@
+ * limitations under the License.
+ */
+
+package org.apache.kafka.raft.internals;
```
You can also move this "composite" metrics to `core`. If you do that then you can specify the broker server metrics and controller server metrics concretely and remove the need for them to "implement" `ExternalKRaftMetricIgnoredStaticVoters`.

kevin-wu24 replied:
Moving this metrics object to `core` results in the following circular dependency when attempting to add this dependency to the `raft` module: 

`Adding dependency on module 'kafka.core.main' will introduce circular dependency between modules 'kafka.coordinator-common.main' and 'kafka.core.main'.`

kevin-wu24 replied:
Maybe this needs to go in `/core/src/main/scala/kafka/raft` specifically.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/ExternalKRaftMetrics.java:
```diff
@@ -0,0 +1,35 @@
+
+import java.util.Optional;
+
+public class ExternalKRaftMetrics {
```
If you move this to `core` and use this name for the interface, then you can call this type something like `DefaultExternalKRaftMetrics`.

jsancio replied:
Let's add a unitest suite for this type. E.g. ExternalKRaftMetricsTest

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KRaftControlRecordStateMachine.java:
```diff
@@ -87,14 +91,23 @@ public KRaftControlRecordStateMachine(
         this.logger = logContext.logger(this.getClass());
+        this.kafkaRaftMetrics = kafkaRaftMetrics;
+        this.externalKRaftMetrics = externalKRaftMetrics;
+        this.staticVoterSet = staticVoterSet != null ? Optional.of(staticVoterSet) : Optional.empty();
```
You can reuse `Optional.ofNullable`.

kevin-wu24 replied:
I need to look at this some more tomorrow, in particular the actual implementation in `KafkaRaftManager`, but `staticVoterSet` might never be `null`. It is never set to `null` in `RaftClientTestContext`, and instead is an empty map when we are running in KIP-853 mode.

IIUC, the intention is for us to only set the metric value if `controller.quorum.voters` is set, but we are not using it. In the case where it isn't set, it looks like `staticVoterSet` is not `null`, so updating the metric in `handleBatch` is still occurring.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KRaftControlRecordStateMachine.java:
```diff
@@ -87,14 +91,23 @@ public KRaftControlRecordStateMachine(
+        this.externalKRaftMetrics = externalKRaftMetrics;
+        this.staticVoterSet = staticVoterSet != null ? Optional.of(staticVoterSet) : Optional.empty();
+
+        if (staticVoterSet != null) {
```
Why are you checking for `null`? If you follow all code paths `staticVoterSet` is never `null`. Did you mean to use `staticVoterSet.ifPresent(voters -> kafkaRaftMetrics.updateNumVoters(voters.size())`?

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KRaftControlRecordStateMachine.java:
```diff
@@ -281,6 +294,11 @@ private void handleBatch(Batch<?> batch, OptionalLong overrideOffset) {
                     VoterSet voters = VoterSet.fromVotersRecord((VotersRecord) record.message());
+                    kafkaRaftMetrics.updateNumVoters(voters.size());
+                    if (staticVoterSet.isPresent() && !ignoredStaticVoterSet) {
+                        ignoredStaticVoterSet = true;
```
I wouldn't bother with this `ignoredStaticVoterSet` optimization. Voters are not changed very often and an atomic or volatile set it not that slow.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/ExternalKRaftMetricIgnoredStaticVoters.java:
```diff
@@ -0,0 +1,22 @@
+
+package org.apache.kafka.raft.internals;
+
+public interface ExternalKRaftMetricIgnoredStaticVoters {
```
Let's write Java docs for the type and method. It is important to state that implementations of this object are no allowed to block.

jsancio commented on metadata/src/main/java/org/apache/kafka/controller/metrics/ControllerMetadataMetrics.java:
```diff
@@ -243,6 +257,13 @@ public void updateUncleanLeaderElection(int count) {
+        ignoredStaticVoters.compareAndSet(false, true);
+    }
+    public boolean ignoredStaticVoters() {
+        return this.ignoredStaticVoters.get();
```
The `this` keyword is not needed.

jsancio commented on metadata/src/main/java/org/apache/kafka/controller/metrics/ControllerMetadataMetrics.java:
```diff
@@ -243,6 +257,13 @@ public void updateUncleanLeaderElection(int count) {
     }
 
+    public void switchIgnoredStaticVoters() {
+        ignoredStaticVoters.compareAndSet(false, true);
```
No need to compare and set. This is the same as just setting to true.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/ExternalKRaftMetricIgnoredStaticVoters.java:
```diff
@@ -0,0 +1,22 @@
+package org.apache.kafka.raft.internals;
+
+public interface ExternalKRaftMetricIgnoredStaticVoters {
+    void switchIgnoredStaticVoters();
```
I know that I suggested using the "switch" prefix but I am starting to think that it is not fully accurate. Maybe the prefix "set" is the best option.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KafkaRaftMetrics.java:
```diff
@@ -208,13 +230,35 @@ public void updateElectionStartMs(long currentTimeMs) {
+    }
+
     @Override
     public void close() {
```
Did you test what happens if `close` gets called before `initialize`?

kevin-wu24 replied:
Calls to `removeMetric` only do something when the metric is actually present in the metrics map, so any metrics that would be added in `initialize` using the `QuorumState` will pass through during `close` in this scenario.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KafkaRaftMetrics.java:
```diff
@@ -208,13 +230,35 @@ public void updateElectionStartMs(long currentTimeMs) {
+        metrics.addMetric(uncommittedVoterChangeMetricName, (Gauge<Integer>) (config, now) -> uncommittedVoterChange);
+    }
+
+    public void removeLeaderMetrics() {
+        metrics.removeMetric(numObserversMetricName);
+        metrics.removeMetric(uncommittedVoterChangeMetricName);
+    }
```
Don't you need to reset the value of those metrics after losing leadership?

kevin-wu24 replied:
Yes, now that we aren't directly reading from LeaderState fields, the local `KafkaRaftMetrics` values need to be reset when leadership is lost.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -682,13 +690,15 @@ private void transitionToCandidate(long currentTimeMs) {
         fetchPurgatory.completeAllExceptionally(
             Errors.NOT_LEADER_OR_FOLLOWER.exception("Not handling request since this node is resigning"));
         quorum.transitionToResigned(preferredSuccessors);
+        kafkaRaftMetrics.removeLeaderMetrics();
```
Okay. Need to check that these are all of the possible transitions from the leader state.

kevin-wu24 replied:
~I think the call below in `onBecomeFollower` is not necessary. Leader state can only transition to Unattached or Resigned~
Never mind, looks like leader -> follower transition with a higher epoch bump is possible during partitioning scenarios (i.e. RaftEventSimulationTest#canMakeProgressAfterBackToBackLeaderFailures). Will add this to the transition chart in `QuorumState`.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/LeaderState.java:
```diff
@@ -217,6 +221,12 @@ public void resetAddVoterHandlerState(
                 .complete(RaftUtil.addVoterResponse(error, message))
         );
         addVoterHandlerState = state;
+        if (addVoterHandlerState.isPresent() || removeVoterHandlerState.isPresent()) {
+            kafkaRaftMetrics.updateUncommittedVoterChange(1);
+        } else {
+            kafkaRaftMetrics.updateUncommittedVoterChange(0);
+        }
```
Code duplication. Let's reuse this code by moving this to a private method.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/LeaderState.java:
```diff
@@ -632,7 +647,10 @@ public long epochStartOffset() {
         ReplicaState state = voterStates.get(replicaKey.id());
         if (state == null || !state.matchesKey(replicaKey)) {
-            observerStates.putIfAbsent(replicaKey, new ReplicaState(replicaKey, false, Endpoints.empty()));
+            LeaderState.ReplicaState previous = observerStates.putIfAbsent(replicaKey, new ReplicaState(replicaKey, false, Endpoints.empty()));
+            if (previous == null) {
+                kafkaRaftMetrics.updateNumObservers(observerStates.size());
+            }
```
Hmm. I am thinking that this optimization is not necessary at the moment. This applies to rest of the changes in this file.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/LeaderState.java:
```diff
@@ -234,6 +244,11 @@ public void resetRemoveVoterHandlerState(
                 .complete(RaftUtil.removeVoterResponse(error, message))
         );
         removeVoterHandlerState = state;
+        if (addVoterHandlerState.isPresent() || removeVoterHandlerState.isPresent()) {
+            kafkaRaftMetrics.updateUncommittedVoterChange(1);
+        } else {
+            kafkaRaftMetrics.updateUncommittedVoterChange(0);
+        }
```
How about
```java
        kafkaRaftMetrics.updateUncommittedVoterChange(
            addVoterHandlerState.isPresent() || removeVoterHandlerState.isPresent()
        );
```

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KafkaRaftMetrics.java:
```diff
@@ -42,6 +42,9 @@ public class KafkaRaftMetrics implements AutoCloseable {
     private volatile OptionalLong pollStartMs;
+    private volatile int numVoters;
+    private volatile int numObservers;
+    private volatile int uncommittedVoterChange;
```
Let's make this a boolean.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KafkaRaftMetrics.java:
```diff
@@ -208,13 +230,35 @@ public void updateElectionStartMs(long currentTimeMs) {
     }
 
+    public void addLeaderMetrics() {
+        metrics.addMetric(numObserversMetricName, (Gauge<Integer>) (config, now) -> numObservers);
+        metrics.addMetric(uncommittedVoterChangeMetricName, (Gauge<Integer>) (config, now) -> uncommittedVoterChange);
```
Do you need the explicit cast to `Gauge<Integer>` in both of this cases? If so, why?

kevin-wu24 replied:
Maybe not necessary but LMK what you think. If I don't include the cast, the second parameter for `addMetric` defaults to `Measurable`, meaning our volatile field serving as the metric value is up-casted to `double` when it doesn't need to be. The metrics introduced in this KIP would always have values of natural numbers ~(maybe `Gauge<Long>` actually most appropriate)~.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java:
```diff
@@ -3850,6 +3851,8 @@ public void testMetrics(boolean withKip853Rpc) throws Exception {
         assertEquals((double) 1L, getMetric(context.metrics, "high-watermark").metricValue());
         assertEquals((double) 1L, getMetric(context.metrics, "log-end-offset").metricValue());
         assertEquals((double) epoch, getMetric(context.metrics, "log-end-epoch").metricValue());
+        assertNotNull(getMetric(context.metrics, "number-of-observers"));
+        assertNotNull(getMetric(context.metrics, "uncommitted-voter-change"));
```
Where it the test for not leader?

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java:
```diff
@@ -3842,6 +3842,7 @@ public void testMetrics(boolean withKip853Rpc) throws Exception {
         assertNotNull(getMetric(context.metrics, "election-latency-max"));
         assertNotNull(getMetric(context.metrics, "fetch-records-rate"));
         assertNotNull(getMetric(context.metrics, "append-records-rate"));
+        assertNotNull(getMetric(context.metrics, "number-of-voters"));
```
We should add a test for the leader specific metric the replica transition from leader to follower to leader. We need to check that after those transitions the metrics are reporting the correct value and not the previous value.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/LeaderStateTest.java:
```diff
@@ -17,9 +17,11 @@
 package org.apache.kafka.raft;
 
 import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.metrics.Metrics;
```
Do we need any unittest test for the metric values?

kevin-wu24 replied:
Adding unit tests here would require us to expose the `kafkaRaftMetrics` object we pass in. After adding the `KafkaRaftClientReconfigTest#testLeaderMetricsAreReset` as well as `KafkaMetricsTest#testLeaderMetrics`, I can't think of any additional cases that are not covered in either that need to be covered here.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java:
```diff
@@ -430,7 +431,8 @@ public RaftClientTestContext build() throws IOException {
                 quorumStateStore,
-                metrics
+                metrics,
+                new ExternalKRaftMetrics(null, null)
```
Where are the test for this metrics like you did for the new metrics in `KafkaRaftMetrics`?

jsancio commented on raft/src/test/java/org/apache/kafka/raft/internals/KRaftControlRecordStateMachineTest.java:
```diff
@@ -19,6 +19,7 @@
 import org.apache.kafka.common.TopicPartition;
 import org.apache.kafka.common.Uuid;
 import org.apache.kafka.common.message.KRaftVersionRecord;
+import org.apache.kafka.common.metrics.Metrics;
```
Where are the unittest for the metric values?

jsancio commented on raft/src/test/java/org/apache/kafka/raft/internals/KafkaRaftMetricsTest.java:
```diff
@@ -92,7 +92,8 @@ private QuorumState buildQuorumState(VoterSet voterSet, KRaftVersion kraftVersio
             new LogContext("kafka-raft-metrics-test"),
-            random
+            random,
+            raftMetrics
```
Don't you need test for the new functionality you added?

jsancio commented on server/src/main/java/org/apache/kafka/server/metrics/BrokerServerMetrics.java:
```diff
@@ -17,19 +17,22 @@
 import static java.util.concurrent.TimeUnit.NANOSECONDS;
 
-public final class BrokerServerMetrics implements AutoCloseable {
+public final class BrokerServerMetrics implements AutoCloseable, ExternalKRaftMetricIgnoredStaticVoters {
```
See my other comments for a suggestion on how to remove this implementation.

jsancio reviewed (commented):
Thanks for your improvements @kevin-wu24

kevin-wu24 committed:
code review p1

kevin-wu24 committed:
code review: adding tests

kevin-wu24 committed:
cleanup LeaderState

kevin-wu24 commented:
Pushed some commits. Changes of note:

- Moved `DefaultExternalKRaftMetrics` to `/core/src/main/scala/kafka/raft` to avoid circular dependency and avoid `BrokerServerMetrics` and `ControllerMetadataMetrics` needing to implement another interface.
- I'm not sure whether or not it makes sense to provide a mock implementation of `ExternalKRaftMetrics` for the purpose of testing. The most recent changes just use the `DefaultExternalKRaftMetrics` implementation directly, since we want to validate that the metrics values in the `brokerServerMetrics` and `controllerMetadataMetrics` objects contained in the implementation are actually being updated. This has the side effect of requiring dependencies being added in the build file for `raft` tests and allowing the relevant imports.
- Added tests in `KafkaMetricsTest` for the newly-added `number-of-voters`, `number-of-observers`, and `uncommitted-voter-change`.
- Modified tests in `KRaftControlRecordStateMachineTest` to test the two `ignored-static-voters` metrics.
- Added `testLeaderMetricsAreReset` in `KafkaRaftClientReconfigTest` to cover the case where a leader loses leadership and regains it, checking that `number-of-observers` and `uncommitted-voter-change` are not using the values from previous leadership.

kevin-wu24 committed:
fixing build

kevin-wu24 committed:
merging trunk

kevin-wu24 committed:
fixing test and cleanup

kevin-wu24 committed:
adding back removal of metrics when transitioning from leader -> follower, updated QuorumState flowchart

jsancio commented on build.gradle:
```diff
@@ -2040,8 +2040,12 @@ project(':raft') {
     testImplementation project(':clients').sourceSets.test.output
+    testImplementation project(':core')
+    testImplementation project(':server')
+    testImplementation project(':metadata')
```
Remove these dependencies. The entire raft module, including tests, should only depend on server-common and clients. The raft module should not depend on core, server or metadata.

jsancio commented on checkstyle/import-control.xml:
Please undo all of the changes to this file. See my other comments on how to do this.

jsancio commented on core/src/main/scala/kafka/raft/DefaultExternalKRaftMetrics.scala:
```diff
@@ -0,0 +1,35 @@
+  val controllerMetadataMetrics: ControllerMetadataMetrics
+) extends ExternalKRaftMetrics {
+  val brokerServerMetricsOpt: Option[BrokerServerMetrics] = Option(brokerServerMetrics)
+  val controllerMetadataMetricsOpt: Option[ControllerMetadataMetrics] = Option(controllerMetadataMetrics)
```
Did you consider making the object parameters optional? We should avoid using `null` in public interfaces. 
```scala
class DefaultExternalKRaftMetrics(
  val brokerServerMetrics: Option[BrokerServerMetrics],
  val controllerMetadataMetrics: Option[ControllerMetadataMetrics]
) extends ExternalKRaftMetrics {
```

https://www.infoq.com/presentations/Null-References-The-Billion-Dollar-Mistake-Tony-Hoare/

kevin-wu24 replied:
I will make the object params optional. Thanks for the presentation link about the reasoning behind this!

jsancio commented on core/src/main/scala/kafka/raft/RaftManager.scala:
```diff
@@ -119,7 +119,8 @@ class KafkaRaftManager[T](
   localListeners: Endpoints,
-  fatalFaultHandler: FaultHandler
+  fatalFaultHandler: FaultHandler,
+  externalKRaftMetrics: ExternalKRaftMetrics
```
I would group this with the `metrics: Metrics` parameter above.

jsancio commented on core/src/main/scala/kafka/tools/TestRaftServer.scala:
```diff
@@ -106,7 +111,11 @@ class TestRaftServer(
+      new DefaultExternalKRaftMetrics(
+        new BrokerServerMetrics(metrics),
+        new ControllerMetadataMetrics(Optional.of(new MetricsRegistry()))
+      )
```
Since these are optional and this is a test raft server, you can just do
```scala
      new DefaultExternalKRaftMetrics(None, None))
```
or this if you keep the old signature.
```scala
      new DefaultExternalKRaftMetrics(null, null))
```

jsancio commented on raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java:
```diff
@@ -649,6 +655,7 @@ private void onBecomeLeader(long currentTimeMs) {
 
         resetConnections();
         kafkaRaftMetrics.maybeUpdateElectionLatency(currentTimeMs);
+        kafkaRaftMetrics.addLeaderMetrics();
```
Did you consider adding the leader metrics when constructing the `LeaderState` object?

LeaderState also implement `void close();`, you can use that invocation to remove any leader specific metrics. The `close` method is called anytime the replica transitions out of a state. This should allow to remove all of the `removeLeaderMetrics` litter through this file.

kevin-wu24 replied:
I didn't consider that. I agree it is better to remove the leader specific metrics in `LeaderState.close()` instead.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/LeaderState.java:
```diff
@@ -237,6 +244,9 @@ public void resetRemoveVoterHandlerState(
                 .complete(RaftUtil.removeVoterResponse(error, message))
         );
         removeVoterHandlerState = state;
+        kafkaRaftMetrics.updateUncommittedVoterChange(
+            addVoterHandlerState.isPresent() || removeVoterHandlerState.isPresent()
+        );
```
Looks like code duplication. Move this to a private method.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/LeaderState.java:
```diff
@@ -691,6 +704,7 @@ private void updateVoterAndObserverStates(VoterSet lastVoterSet) {
         for (ReplicaState replicaStateEntry : oldVoterStates.values()) {
             replicaStateEntry.clearListeners();
             observerStates.putIfAbsent(replicaStateEntry.replicaKey, replicaStateEntry);
+            kafkaRaftMetrics.updateNumObservers(observerStates.size());
```
This method updates the number of observers N times. How about doing it once after all of the changes to `observerStates` are done?

jsancio commented on raft/src/main/java/org/apache/kafka/raft/QuorumState.java:
```diff
@@ -95,6 +97,8 @@ public class QuorumState {
     private final int fetchTimeoutMs;
     private final LogContext logContext;
 
+    private final KafkaRaftMetrics kafkaRaftMetrics;
```
Extra space if you want to keep it consistent with the rest of the private fields.

jsancio commented on metadata/src/main/java/org/apache/kafka/controller/metrics/ControllerMetadataMetrics.java:
```diff
@@ -65,6 +69,8 @@ public final class ControllerMetadataMetrics implements AutoCloseable {
     private final AtomicInteger metadataErrorCount = new AtomicInteger(0);
     private Optional<Meter> uncleanLeaderElectionMeter = Optional.empty();
 
+    private final AtomicBoolean ignoredStaticVoters = new AtomicBoolean(false);
```
Extra space if you want to keep it consistent with the rest of the private fields.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KRaftControlRecordStateMachine.java:
```diff
@@ -87,14 +91,21 @@ public KRaftControlRecordStateMachine(
+        this.externalKRaftMetrics = externalKRaftMetrics;
+        this.staticVoterSet = Optional.ofNullable(staticVoterSet);
+
+        this.staticVoterSet.ifPresent(voters -> kafkaRaftMetrics.updateNumVoters(voters.size()));
```
Take a look at `truncateNewEntries`. The number of voters may also get updated because of log end truncation.

`truncateOldEntries` should never truncate the latest entry but please take a look at that code to confirm.

kevin-wu24 replied:
`truncateOldEntries` never truncates the latest entry: 
```
@Override
    public void truncateOldEntries(long startOffset) {
        NavigableMap<Long, T> lesserValues = history.headMap(startOffset, true);
        while (lesserValues.size() > 1) {
            // Poll and ignore the entry to remove the first entry
            lesserValues.pollFirstEntry();
        }
    }
```

kevin-wu24 replied:
If `truncateNewEntries` removes the oldest voter set from the log, would we go back to the static configuration if it exists (we would need to update `ignored-static-voters` in that case, no)? What if it doesn't exist?

jsancio replied:
Yes. that's correct. In that case the value of `ignored-static-voters` will go from true to false.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/VoterSet.java:
```diff
@@ -158,6 +158,13 @@ public Set<VoterNode> voterNodes() {
     }
 
+    /**
+     * Returns size of the voter set.
```
How about "Returns the number of voters in the voter set."?

jsancio commented on server/src/main/java/org/apache/kafka/server/metrics/BrokerServerMetrics.java:
```diff
@@ -55,13 +57,17 @@ public final class BrokerServerMetrics implements AutoCloseable {
     private final AtomicLong metadataLoadErrorCount = new AtomicLong(0);
     private final AtomicLong metadataApplyErrorCount = new AtomicLong(0);
 
+    private final AtomicBoolean ignoredStaticVoters = new AtomicBoolean(false);
```
You can remove the extra newline.

jsancio commented on server/src/main/java/org/apache/kafka/server/metrics/BrokerServerMetrics.java:
```diff
@@ -55,13 +57,17 @@ public final class BrokerServerMetrics implements AutoCloseable {
     private final MetricName metadataLoadErrorCountName;
     private final MetricName metadataApplyErrorCountName;
 
+    private final MetricName ignoredStaticVotersName;
```
You can remove the extra newline.

jsancio commented on server/src/main/java/org/apache/kafka/server/metrics/BrokerServerMetrics.java:
```diff
@@ -89,12 +95,18 @@ public BrokerServerMetrics(Metrics metrics) {
+        ignoredStaticVotersName = metrics.metricName(
+            "ignored-static-voters",
+            METRIC_GROUP_NAME,
+            "This value is 1 when the current voter set for the metadata topic partition is being read from the log and 0 when it is read from the static configuration."
```
This description doesn't seem accurate. Is this what I state in the KIP? How about:

> 1 if controller.quorum.voters is set but was not used by the broker, 0 otherwise.

Please feel free to update the KIP.

kevin-wu24 replied:
Will do, requested permissions to contribute via the developer mailing list. Also will update the KIP with why we're not implementing `number-of-offline-voters` and `is-observer`.

jsancio reviewed (commented):
Thanks for the changes. I reviewed `src/main`.

kevin-wu24 committed:
code review

kevin-wu24 committed:
Merge branch 'trunk' into KAFKA-16524

kevin-wu24 committed:
covering case where ignored-static-voters should be set to false when we fallback to static voter config

kevin-wu24 committed:
can check voter history last entry DNE instead of comparing voter sets to determine if truncation results in using static voters config

kevin-wu24 commented:
Thanks for the review @jsancio. Pushed some commits:

- Added `MockExternalKRaftMetrics` to prevent additional raft module dependencies. Added `DefaultExternalKRaftMetricsTest` since raft layer tests use the mock instead.
- The `staticVoterSet` passed to `KRaftControlRecordStateMachine` can never be `null`. Instead the default value for `controller.quorum.voters` is the empty list in both implementation and testing. Fixed updating logic for the `ignored-static-voters` metric so that it always stays false when the mentioned config is not set. Only when that config is actually set can the metric's value ever go from false to true.
- In the case of `truncateNewEntries` truncating the entire log, if `controller.quorum.voters` is set, `ignored-static-voters` metric can go from true back to false. Implementation now covers for this case.
- `KRaftControlRecordStateMachineTest` checks both `ignored-static-voters` and `number-of-voters` metric values.

kevin-wu24 committed:
cleanup

jsancio commented on core/src/test/scala/kafka/raft/DefaultExternalKRaftMetricsTest.scala:
```diff
@@ -38,17 +38,17 @@ final class DefaultExternalKRaftMetricsTest {
 
     metrics = new DefaultExternalKRaftMetrics(None, None)
-    metrics.setIgnoredStaticVoters()
+    metrics.setIgnoredStaticVoters(true)
```
Make this a different test.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KRaftControlRecordStateMachine.java:
```diff
@@ -70,6 +71,9 @@ public final class KRaftControlRecordStateMachine {
     private volatile long nextOffset = STARTING_NEXT_OFFSET;
+    private final KafkaRaftMetrics kafkaRaftMetrics;
+    private final ExternalKRaftMetrics externalKRaftMetrics;
+    private final Optional<VoterSet> staticVoterSet;
```
You can just model this with `VoterSet staticVoterSet`. `Optional.empty()` has the same meaning as `VoterSet.empty()` or `VoterSet#isEmpty()`.

jsancio commented on core/src/main/scala/kafka/raft/DefaultExternalKRaftMetrics.scala:
```diff
@@ -0,0 +1,33 @@
+import org.apache.kafka.server.metrics.BrokerServerMetrics
+
+class DefaultExternalKRaftMetrics(
+  val brokerServerMetricsOpt: Option[BrokerServerMetrics],
+  val controllerMetadataMetricsOpt: Option[ControllerMetadataMetrics]
```
Okay but in general Kafka doesn't tend to encode the type (Option) of the variable in the variable name. E.g. you can call it `controllerMetadataMetrics` instead of `controllerMetadataMetricsOpt`.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/QuorumState.java:
```diff
@@ -722,8 +727,11 @@ public <T> LeaderState<T> transitionToLeader(long epochStartOffset, BatchAccumul
+            logContext,
+            kafkaRaftMetrics
         );
+        kafkaRaftMetrics.addLeaderMetrics();
```
Since `LeaderState#close` removes these metrics you should keep this coupling in the `LeaderState` constructor. In other words move `kafkaRaftMetrics.addLeaderMetrics()` to the constructor for `LeaderState`. This is also called resource acquisition is initialization (RAII).

kevin-wu24 replied:
Ah yeah, good catch. I remember RAII being mentioned from some C++ I did in college.

jsancio reviewed (commented):
Thanks. Submitted a quick review.

kevin-wu24 committed:
code review

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KRaftControlRecordStateMachine.java:
```diff
@@ -103,9 +103,11 @@ public KRaftControlRecordStateMachine(
+        this.staticVoterSet = staticVoterSet;
 
-        this.staticVoterSet.ifPresent(voters -> kafkaRaftMetrics.updateNumVoters(voters.size()));
+        if (staticVoterSet.size() > 0) {
+            kafkaRaftMetrics.updateNumVoters(staticVoterSet.size());
+        }
```
```java
        kafkaRaftMetrics.updateNumVoters(staticVoterSet.size());
```
Do you need the conditional?

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KRaftControlRecordStateMachine.java:
```diff
@@ -130,7 +132,7 @@ public void truncateNewEntries(long endOffset) {
 
         kafkaRaftMetrics.updateNumVoters(voterSetHistory.lastValue().size());
-        if (staticVoterSet.isPresent() && voterSetHistory.lastEntry().isEmpty()) {
+        if (staticVoterSet.size() > 0 && voterSetHistory.lastEntry().isEmpty()) {
```
Use `!staticVoterSet.isEmpty()`. Sometimes `size()` is not constant time.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KRaftControlRecordStateMachine.java:
```diff
@@ -298,7 +300,7 @@ private void handleBatch(Batch<?> batch, OptionalLong overrideOffset) {
                     VoterSet voters = VoterSet.fromVotersRecord((VotersRecord) record.message());
                     kafkaRaftMetrics.updateNumVoters(voters.size());
-                    if (staticVoterSet.isPresent()) {
+                    if (staticVoterSet.size() > 0) {
```
Use `!staticVoterSet.isEmpty()`. Sometimes `size()` is not constant time.

kevin-wu24 committed:
code review

kevin-wu24 committed:
fixing metrics tests

jsancio commented on metadata/src/main/java/org/apache/kafka/controller/metrics/ControllerMetadataMetrics.java:
```diff
@@ -117,6 +120,13 @@ public Integer value() {
         registry.ifPresent(r -> uncleanLeaderElectionMeter =
                 Optional.of(registry.get().newMeter(UNCLEAN_LEADER_ELECTIONS_PER_SEC, "elections", TimeUnit.SECONDS)));
+
+        registry.ifPresent(r -> r.newGauge(IGNORED_STATIC_VOTERS, new Gauge<Boolean>() {
```
I took at look at this using the MBeans browser. Using a boolean doesn't expose the value as 1 for true and 0 for false. You need to expose the gauge as an integer: `return ignoredStaticVoter() ? 1 : 0`

jsancio commented on metadata/src/test/java/org/apache/kafka/controller/metrics/ControllerMetadataMetricsTest.java:
```diff
@@ -48,6 +48,7 @@ public void testMetricNames() {
                         "kafka.controller:type=KafkaController,name=MetadataErrorCount",
                         "kafka.controller:type=KafkaController,name=OfflinePartitionsCount",
                         "kafka.controller:type=KafkaController,name=PreferredReplicaImbalanceCount",
+                        "kafka.controller:type=KafkaController,name=IgnoredStaticVoters",
```
Let's add a test specific for this metric that checks that the value is 1 or 0 when expected.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientReconfigTest.java:
```diff
@@ -2351,4 +2459,21 @@ private static ApiVersionsResponseData apiVersionsResponse(Errors error, Support
+        final int expectedVoters,
+        final int expectedObservers,
+        final boolean expectedUncommittedVoterChange,
+        final boolean expectedIgnoredStaticVoters,
```
I couldn't find a single test in this suite that check that this value (`expectedIgnoredStaticVoters`) is true.

kevin-wu24 replied:
I can add a test where we:
1. set bootstrap records which define the voter set via a snapshot
2. explicitly set `isStartingVotersStatic` to be true so the test's RaftClient has static voters

This test does expect this metric value to be `true` after the client is set up, but the snapshot voter set and the RaftClient's static voters are the same set because of how we implement `RaftClientTestContext`.

kevin-wu24 replied:
I guess my main question is about the necessity to test this the `ignored-static-voters` metric in tests that use `RaftClientTestContext`. Because of the way we initialize the context, this metric will either report true or false for the entirety of the test. I feel this is not very useful to test, and that is why I added tests for this metric to `KRaftControlRecordStateMachineTest`.

My thinking is after version upgrades are implemented I think checking the metric value in these tests makes more sense (i.e. a case where we initialize the context in version zero, upgrade to version 1, check the metric is still false, update the voter set, finally check the metric is true).

jsancio replied:
Makes sense to me to not test this configuration here. In that case don't make it a parameter and simply check for false in the method body.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/MockExternalKRaftMetrics.java:
```diff
@@ -0,0 +1,35 @@
+ * This class is used to replicate the behavior of {@link kafka.raft.DefaultExternalKRaftMetrics}
+ * for testing within the raft layer.
+ */
+public class MockExternalKRaftMetrics implements ExternalKRaftMetrics {
```
This seems to be a very simple mock. Did you try using Mockito's mock? E.g. `Mockito.mock`. If you search for that method, you should find many examples of using that method.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/internals/KRaftControlRecordStateMachineTest.java:
```diff
@@ -51,10 +58,28 @@ private static KRaftControlRecordStateMachine buildPartitionListener(MockLog log
+        metrics = new Metrics();
+        kafkaRaftMetrics = new KafkaRaftMetrics(metrics, "raft");
+        externalMetrics = new MockExternalKRaftMetrics();
+    }
```
Please void using non-final static or instance variables, and resetting them using `BeforeEach` and `AfterEach` methods. They make tests very difficult to read and maintain.

Please create these objects explicitly in every test that needs them.

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KafkaRaftMetrics.java:
```diff
@@ -137,6 +99,14 @@ public KafkaRaftMetrics(Metrics metrics, String metricGrpPrefix, QuorumState sta
         metrics.addMetric(this.numUnknownVoterConnectionsMetricName, (mConfig, currentTimeMs) -> numUnknownVoterConnections);
 
+        this.numVotersMetricName = metrics.metricName("number-of-voters", metricGroupName, "Number of voters for a KRaft topic partition.");
+        metrics.addMetric(this.numVotersMetricName, (Gauge<Integer>) (mConfig, currentTimestamp) -> numVoters);
```
I still don't understand why you need this cast to `Gauge<Integer>`. The metric above `numUnknownVoterConnectionsMetricName` is also an integer and it doesn't need a cast. Why is this behaving differently?

kevin-wu24 replied:
Maybe I'm misunderstanding. Whenever I don't perform an explicit cast to a `Gauge<T>`, the method being called is 
```
addMetric(MetricName metricName, Measurable measurable)
```
and not
```
public void addMetric(MetricName metricName, MetricValueProvider<?> metricValueProvider)
```
When the passed in lambda is an implementation of `Measurable`, calling `measure()` to read the metric value returns a `double`, meaning `int numVoters` gets up-casted to a `double` when it doesn't need to be. 

This applies to other metrics as well, not just the one added in this PR.

jsancio replied:
I understand now. Does that mean that some of the existing metrics are wrong? They expose a double instead of an integer?

Can you file a jira documenting the metrics in this object that expose a double when it should be an integer? Can you share the link to the jira here?

kevin-wu24 replied:
I believe we are exposing doubles unintentionally. When we take the metric and return the value in `KafkaMetric.metricValue()`:
```
@Override
    public Object metricValue() {
        long now = time.milliseconds();
        synchronized (this.lock) {
            if (isMeasurable())
                return ((Measurable) metricValueProvider).measure(config, now);
            else if (this.metricValueProvider instanceof Gauge)
                return ((Gauge<?>) metricValueProvider).value(config, now);
            else
                throw new IllegalStateException("Not a valid metric: " + this.metricValueProvider.getClass());
        }
    }
```
We would be calling `Measurable.measure()` and returning a double.

I don't think I've ever seen this kind of case before where we overload a method signature in this fashion (the first signature takes in `Measurable`, an interface that extends MetricValueProvider<Double>, while the second signature takes a generic of the `MetricValueProvider<T>` parent interface). 

Unless I perform an explicit cast/declaration of a Gauge<Integer/Boolean/etc.> object to pass in as the second parameter of `addMetric`, the compiler always assigns the method to call as `addMetric(MetricName, Measurable)`. I think this is because the second method signature is taking in the super-interface `MetricValueProvider` as a parameter, which is not a functional interface and cannot be implemented with a lambda like `Measurable`, rather than its child `Gauge<T>`.

kevin-wu24 replied:
Link to jira: https://issues.apache.org/jira/browse/KAFKA-18679

jsancio commented on raft/src/main/java/org/apache/kafka/raft/internals/KafkaRaftMetrics.java:
```diff
@@ -207,13 +229,37 @@ public void updateElectionStartMs(long currentTimeMs) {
 
+    public void addLeaderMetrics() {
+        metrics.addMetric(numObserversMetricName, (Gauge<Integer>) (config, now) -> numObservers);
+        metrics.addMetric(uncommittedVoterChangeMetricName, (Gauge<Boolean>) (config, now) -> uncommittedVoterChange);
```
This exposes the metrics as false/true. The KIP mentions that the metric is meant to expose 0/1.

jsancio commented on raft/src/test/java/org/apache/kafka/raft/internals/KafkaRaftMetricsTest.java:
```diff
@@ -469,6 +478,54 @@ public void shouldRecordRate(KRaftVersion kraftVersion) {
+        raftMetrics.addLeaderMetrics();
+
+        assertEquals(0, getMetric(metrics, "number-of-observers").metricValue());
+        assertEquals(false, getMetric(metrics, "uncommitted-voter-change").metricValue());
```
I made a similar comment in the `src/main` implementation but this metric should be an integer with the values 1 or 0.

jsancio commented on server/src/main/java/org/apache/kafka/server/metrics/BrokerServerMetrics.java:
```diff
@@ -89,12 +93,18 @@ public BrokerServerMetrics(Metrics metrics) {
         metrics.addMetric(lastAppliedRecordLagMsName, (config, now) -> now - lastAppliedImageProvenance.get().lastContainedLogTimeMs());
         metrics.addMetric(metadataLoadErrorCountName, (config, now) -> metadataLoadErrorCount.get());
         metrics.addMetric(metadataApplyErrorCountName, (config, now) -> metadataApplyErrorCount.get());
+        metrics.addMetric(ignoredStaticVotersName, (Gauge<Boolean>) (config, now) -> ignoredStaticVoters.get());
```
According to the KIP is metrics is expected to have the values 1 and 0, not true and false.

jsancio commented on server/src/test/java/org/apache/kafka/server/metrics/BrokerServerMetricsTest.java:
```diff
@@ -47,7 +47,8 @@ public void testMetricsExported() throws Exception {
                 new MetricName("metadata-load-error-count", expectedGroup, "", Collections.emptyMap()),
-                new MetricName("metadata-apply-error-count", expectedGroup, "", Collections.emptyMap())
+                new MetricName("metadata-apply-error-count", expectedGroup, "", Collections.emptyMap()),
+                new MetricName("ignored-static-voters", expectedGroup, "", Collections.emptyMap())
```
Can we add a test specific to the new metrics "ignored-static-voters"?

jsancio reviewed (commented):
Thanks for the update @kevin-wu24 . Let's make sure that the metrics that represent a boolean expose the value as 1 or 0 and not as true or false.

kevin-wu24 committed:
code review: metrics that represent boolean expose 1 or 0, removing mock, adding tests

kevin-wu24 committed:
fixing build

jsancio reviewed (approved):
LGTM

chia7712 commented:
> Moved DefaultExternalKRaftMetrics to /core/src/main/scala/kafka/raft to avoid circular dependency and avoid BrokerServerMetrics and ControllerMetadataMetrics needing to implement another interface.

Out of curiosity, considering the eventual removal of the 'core' module, what is the planned strategy for resolving the circular dependency in the future?

kevin-wu24 commented:
> Out of curiosity, considering the eventual removal of the 'core' module, what is the planned strategy for resolving the circular dependency in the future?

We'll probably have to do something like what's mentioned here: https://github.com/apache/kafka/pull/18304#discussion_r1902242222, where `ExternalKRaftMetrics` lives in `raft/internals`, along with the interface that `BrokerServerMetrics` and `ControllerMetadataMetrics` implement.
